### PR TITLE
feat: storage node complete recovery

### PR DESCRIPTION
**Pull request by halfprice**
This PR implements a mechanism which brings a node that is lagging for arbitrarily long to the last epoch.

The main approach is to only process BlobInfo update without syncing blobs, and after reaching the latest epoch, the node scans the BlobInfo table and finds all the certified blobs to recover/sync.

### CODE REVIEW

File: crates/walrus-sdk/src/client.rs [hunk 1]
```diff
@@ -172,7 +172,9 @@ impl UrlEndpoints {
 
     fn server_health_info(&self) -> (Url, &'static str) {
         (
-            self.0.join("/v1/health").expect("this is a valid URL"),
+            self.0
+                .join("/v1/health?detailed=true")
+                .expect("this is a valid URL"),
             HEALTH_URL_TEMPLATE,
         )
```
Comment (mlegner) [inline crates/walrus-sdk/src/client.rs]: Suggestion: Add `detailed: bool` as a function argument.

File: crates/walrus-service/src/node/committee.rs [hunk 2]
```diff
@@ -171,6 +171,8 @@ pub trait CommitteeService: std::fmt::Debug + Send + Sync {
     /// Otherwise, an error is returned.
     fn end_committee_change(&self, epoch: Epoch) -> Result<(), EndCommitteeChangeError>;
 
+    async fn update_to_latest_committee(&self) -> Result<(), BeginCommitteeChangeError>;
```
Comment (mlegner) [inline crates/walrus-service/src/node/committee.rs]: It's a bit unclear how this is related to the `begin_committee_change`. If it's something different, it might also warrant its own error type.

File: crates/walrus-service/src/node.rs [hunk 3]
```diff
@@ -568,13 +578,16 @@ impl StorageNode {
                     // Clear the starting epoch, so that we never make this check again.
                     maybe_epoch_at_start = None;
 
-                    // if event.event_epoch() < starting_epoch - 1
-                    if event.event_epoch() + 1 < epoch_at_start {
+                    if node_status == NodeStatus::Active && event.event_epoch() + 1 < epoch_at_start
+                    {
                         tracing::warn!(
-                            "the current epoch ({}) is too far ahead of the event epoch: {}",
+                            "the current epoch ({}) is too far ahead of the event epoch: {}; node entering recovery mode",
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Nit: Given that we have the recovery mode now, we could remove the "too" 😁.
```suggestion
                            "the current epoch ({}) is far ahead of the event epoch: {}; node entering recovery mode",
```

File: crates/walrus-service/src/node.rs [hunk 4]
```diff
@@ -813,6 +837,53 @@ impl StorageNode {
...
+                tracing::info!(
+                    "processing event reaches the latest epoch, start recover entire node"
+                );
+
+                self.inner
+                    .storage
+                    .set_node_status(NodeStatus::RecoveryInProgress)?;
+
+                // Cancel any shard sync;
+                self.shard_sync_handler.cancel_all().await;
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Question: Why don't we want to do shard syncs? The node may have gained a shard right in this epoch, in which case it would be great to sync that shard instead of recovering it.

File: crates/walrus-service/src/node/node_recovery.rs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            task_handle: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    pub fn start_node_recovery(&self, epoch: Epoch) -> Result<(), TypedStoreError> {
+        let start = tokio::time::Instant::now();
+        let mut locked_task_handle = self.task_handle.lock().unwrap();
+        assert!(locked_task_handle.is_none());
+
+        let node: Arc<StorageNodeInner> = self.node.clone();
```
Comment (mlegner) [inline crates/walrus-service/src/node/node_recovery.rs]: Nit: Is the type here necessary?
```suggestion
        let node = self.node.clone();
```

File: crates/walrus-service/src/node/node_recovery.rs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                    match blob_result {
+                        Ok((blob_id, blob_info)) => {
+                            if !blob_info.is_certified(epoch) {
+                                // Skip blobs that are not certified in the given epoch. This includes blobs that are
+                                // invalid or expired.
+                                tracing::info!(
+                                    blob_id = ?blob_id,
+                                    epoch = epoch,
+                                    "skip non-certified blob"
+                                );
```
Comment (mlegner) [inline crates/walrus-service/src/node/node_recovery.rs]: Nit: This should probably be `debug!` as there may be many.

File: crates/walrus-service/src/node/node_recovery.rs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                                continue;
+                            }
+
+                            if let Ok(stored_at_all_shards) =
+                                node.storage.is_stored_at_all_shards(&blob_id)
+                            {
+                                if stored_at_all_shards {
```
Comment (mlegner) [inline crates/walrus-service/src/node/node_recovery.rs]: Nit: This version is a bit shorter:
```suggestion
                            if let Ok(true) = node.storage.is_stored_at_all_shards(&blob_id) {
```

File: crates/walrus-service/src/node/node_recovery.rs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                let mut contain_recovery_blob = false;
+                for blob_result in node.storage.certified_blob_info_iter_before_epoch(epoch) {
+                    match blob_result {
+                        Ok((blob_id, blob_info)) => {
+                            if !blob_info.is_certified(epoch) {
+                                // Skip blobs that are not certified in the given epoch. This includes blobs that are
+                                // invalid or expired.
+                                tracing::info!(
+                                    blob_id = ?blob_id,
+                                    epoch = epoch,
```
Comment (mlegner) [inline crates/walrus-service/src/node/node_recovery.rs]: Suggestion: Use the `Display` formatter here. Also, the `=` is not necessary. Both also apply to other places.
```suggestion
                                    %blob_id,
                                    epoch,
```

File: crates/walrus-service/src/node/node_recovery.rs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+                        Err(err) => {
+                            tracing::error!(error = ?err, "failed to read certified blob");
+                        }
+                    }
+                }
+
+                if !contain_recovery_blob {
+                    tracing::info!("no recovery blob found; stop recovery task");
+                    break;
+                }
```
Comment (mlegner) [inline crates/walrus-service/src/node/node_recovery.rs]: Question: Do we actually need this boolean flag when we have the `all_blob_syncs` vector?

File: crates/walrus-service/src/node/node_recovery.rs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+                            match start_sync_result {
+                                Ok(notify) => {
+                                    all_blob_syncs.push(notify);
+                                }
+                                Err(err) => {
+                                    tracing::error!(
+                                        blob_id = ?blob_id,
+                                        error = ?err,
+                                        "failed to start recovery sync for blob",
+                                    );
```
Comment (mlegner) [inline crates/walrus-service/src/node/node_recovery.rs]: Question: What do we do with the blobs we fail to recover?

File: crates/walrus-service/src/node/blob_sync.rs [hunk 11]
```diff
@@ -44,6 +48,12 @@ struct Permits {
     sliver: Arc<Semaphore>,
 }
 
+#[derive(Debug, Copy, Clone)]
+pub struct EventInfo {
+    pub event_id: EventID,
+    pub event_index: usize,
+}
```
Comment (mlegner) [inline crates/walrus-service/src/node/blob_sync.rs]: Just FYI, this has some overlap with my PR #1125. 😁

Should be simple to resolve though.

File: crates/walrus-service/src/node/blob_sync.rs
Comment (mlegner) [inline crates/walrus-service/src/node/blob_sync.rs:1]: Question: With these changes to the blob sync, could we also use it for the recovery if shard sync fails? This would then probably solve #1095 automatically. There would still remain the inefficiency of checking all shards instead of focusing on the one(s) that need to be recovered.

File: crates/walrus-service/src/node/storage.rs [hunk 12]
```diff
@@ -52,13 +55,39 @@ pub(crate) use shard::{ShardStatus, ShardStorage};
 #[derive(Debug, Clone, Copy)]
 pub struct WouldBlockError;
 
+// TODO: add status to health endpoint
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage.rs]: Suggestion: Can we then also add a metric for it? (I'm currently adding some metrics in #1127.)

File: crates/walrus-sdk/src/client.rs (hunk 1 above)
Comment (halfprice) [reply crates/walrus-sdk/src/client.rs]: Done

File: crates/walrus-service/src/node.rs (hunk 4 above)
Comment (halfprice) [reply crates/walrus-service/src/node.rs]: I removed this: if there is any existing shard sync on going, we won't cancel it to save some work.

The main reason I wanted to cancel the existing shard sync is for simplicity. Theoretically, if the node gained the shard 2 epoch ago, the original source may not hold the shard anymore, and stop the sync immediately.

File: crates/walrus-service/src/node/node_recovery.rs (hunk 9 above)
Comment (halfprice) [reply crates/walrus-service/src/node/node_recovery.rs]: Good point. Removed

File: crates/walrus-service/src/node/storage.rs (hunk 12 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage.rs]: Done.

File: crates/walrus-service/src/node/blob_sync.rs
Comment (halfprice) [reply crates/walrus-service/src/node/blob_sync.rs:1]: Yes, but probably in the next PR. I agree that one nice thing about reuse blob sync handler is that cancellation is handled.

File: crates/walrus-service/src/node/node_recovery.rs (hunk 10 above)
Comment (halfprice) [reply crates/walrus-service/src/node/node_recovery.rs]: It's actually impossible in this code path. There only place where start_sync can return error is when marking event complete. In this codepath, there is no event to mark, and therefore cannot fail. I added more comments here.

File: crates/walrus-service/src/node/committee.rs (hunk 2 above)
Comment (halfprice) [reply crates/walrus-service/src/node/committee.rs]: The reason I reused the error type is mainly due to the duplicated code path between `update_to_latest_committee` and `begin_committee_change` where they fetch the new committee on chain. I wonder if it make sense to change `update_to_latest_committee` to `begin_committee_change_to_latest_committee` so that reusing the error makes more sense?

File: crates/walrus-service/src/node/committee/committee_service.rs [hunk 13]
```diff
@@ -611,6 +604,57 @@ where
     fn end_committee_change(&self, epoch: Epoch) -> Result<(), EndCommitteeChangeError> {
         self.end_committee_change_to(epoch)
     }
+
+    async fn update_to_latest_committee(&self) -> Result<(), BeginCommitteeChangeError> {
```
Comment (halfprice) [inline crates/walrus-service/src/node/committee/committee_service.rs]: Hi @giac-mysten , can you please take a look at this function and see if it makes sense? I didn't touched the committee updating code path before, so I don't know if I may miss something. The idea in this function is to update the committee in the node to match the latest committee on chain. Different from `begin_committee_change` where it has to move from committee in epoch X to epoch X+1, this function has no restriction on the new committee's epoch: we just want to get the latest on chain.

Thanks!

File: crates/walrus-sdk/src/client.rs [hunk 14]
```diff
@@ -170,11 +170,20 @@ impl UrlEndpoints {
...
+                    .join("/v1/health?detailed=true")
+                    .expect("this is a valid URL"),
+                HEALTH_URL_TEMPLATE,
+            )
+        } else {
+            (
+                self.0.join("/v1/health").expect("this is a valid URL"),
+                HEALTH_URL_TEMPLATE,
+            )
+        }
```
Comment (mlegner) [inline crates/walrus-sdk/src/client.rs]: Nit:
```suggestion
        let mut url = self.0.join("/v1/health").expect("this is a valid URL");
        url.set_query(detailed.then_some("detailed=true"));
        (url, HEALTH_URL_TEMPLATE)
```

File: crates/walrus-service/src/node.rs [hunk 15]
```diff
@@ -568,13 +578,24 @@ impl StorageNode {
...
                             epoch_at_start,
                             event.event_epoch()
                         );
+                        self.inner
+                            .metrics
+                            .current_node_status
+                            .set(NodeStatus::RecoveryCatchUp.to_i64());
+                        self.inner
+                            .storage
+                            .set_node_status(NodeStatus::RecoveryCatchUp)?;
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Nit: Can we create a function `StorageNodeInner::set_node_status` that combines these two? That way we won't forget to update the metrics.

File: crates/walrus-sdk/src/client.rs [hunk 16]
```diff
@@ -956,7 +968,7 @@ mod tests {
     #[test]
     fn test_url_health_info_endpoint() {
         let endpoints = UrlEndpoints(Url::parse("https://node.com").unwrap());
-        let (url, _) = endpoints.server_health_info();
+        let (url, _) = endpoints.server_health_info(false);
 
         assert_eq!(url.to_string(), "https://node.com/v1/health");
     }
```
Comment (mlegner) [inline crates/walrus-sdk/src/client.rs]: Here we could add a test for the detailed version:
```rust


    param_test! {
        test_url_health_info_endpoint: [
            default: (false, "https://node.com/v1/health"),
            detailed: (true, "https://node.com/v1/health?detailed=true"),
        ]
    }
    fn test_url_health_info_endpoint(detailed: bool, expected_url: &str) {
        let endpoints = UrlEndpoints(Url::parse("https://node.com").unwrap());
        let (url, _) = endpoints.server_health_info(detailed);

        assert_eq!(url.to_string(), expected_url);
    }
```

File: crates/walrus-service/src/node.rs [hunk 17]
```diff
@@ -819,6 +877,62 @@ impl StorageNode {
...
+            .metrics
+            .current_node_status
+            .set(NodeStatus::RecoveryInProgress(event.epoch).to_i64());
+        self.inner
+            .storage
+            .set_node_status(NodeStatus::RecoveryInProgress(event.epoch))?;
+
+        // // Cancel any shard syncs that are currently in progress. Since the node is lagging,
+        // self.shard_sync_handler.cancel_all().await;
+
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Nit:
```suggestion
```

File: crates/walrus-service/src/node/blob_sync.rs [hunk 18]
```diff
@@ -170,65 +173,81 @@ impl BlobSyncHandler {
...
                     "otel.kind" = "CONSUMER",
                     "otel.status_code" = field::Empty,
                     "otel.status_message" = field::Empty,
-                    "walrus.event.index" = event_index,
-                    "walrus.event.tx_digest" = ?event.event_id.tx_digest,
-                    "walrus.event.event_seq" = ?event.event_id.event_seq,
-                    "walrus.event.kind" = event.label(),
-                    "walrus.blob_id" = %event.blob_id,
+                    "walrus.event.index" = event_info.map(|info| info.event_index),
+                    "walrus.event.tx_digest" = ?event_info.map(|info| info.event_id),
```
Comment (mlegner) [inline crates/walrus-service/src/node/blob_sync.rs]:
```suggestion
                    "walrus.event.tx_digest" = ?event_info.map(|info| info.event_id.tx_digest),
```

File: crates/walrus-service/src/node/blob_sync.rs [hunk 19]
```diff
@@ -275,8 +300,8 @@ impl BlobSyncHandler {
         self.remove_sync_handle(&synchronizer.blob_id).await;
 
         let label = match output {
-            Ok(Some(_)) => metrics::STATUS_SUCCESS,
-            Ok(None) => metrics::STATUS_CANCELLED,
+            Ok(None) => metrics::STATUS_SUCCESS,
```
Comment (mlegner) [inline crates/walrus-service/src/node/blob_sync.rs]: This could also happen when the sync is cancelled and `synchronizer.event_info.is_none()`, right?

File: crates/walrus-service/src/node/committee.rs (hunk 2 above)
Comment (mlegner) [reply crates/walrus-service/src/node/committee.rs]: That sounds good, yes.

File: crates/walrus-service/src/node/metrics.rs [hunk 20]
```diff
@@ -90,7 +91,10 @@ telemetry::define_metric_set! {
             // Buckets from 2^18 (256 KiB) to 2^34 (16 GiB, inclusive), which covers the ~212 KiB to
             // 14 GiB unencoded blobs which are possible in a system with 1000 shards.
             buckets: (18..=34).map(|power| (1u64 << power) as f64).collect::<Vec<_>>()
-        }
+        },
+
+        #[help = "Indicates the current node status"]
+        current_node_status: IntGauge[],
```
Comment (mlegner) [inline crates/walrus-service/src/node/metrics.rs:97]: Remark: We use a different way of encoding the epoch state, see [here](https://github.com/MystenLabs/walrus/blob/9a31a5e3ab04b99513d8f1c2d5d1f54ef04bd7ae/crates/walrus-service/src/common/telemetry.rs#L448). I think JP tried to mimic the systemd metrics, see https://github.com/MystenLabs/walrus/pull/983. However, your version seems simpler and more efficient to me. 🤷

File: crates/walrus-service/src/node/shard_sync.rs [hunk 21]
```diff
@@ -215,6 +215,14 @@ impl ShardSyncHandler {
...
     }
 
+    // /// Cancels all the shard sync tasks.
+    // pub async fn cancel_all(&self) {
+    //     let mut shard_sync_in_progress = self.shard_sync_in_progress.lock().await;
+    //     for (_, task) in shard_sync_in_progress.drain() {
+    //         task.abort();
+    //     }
+    // }
+
```
Comment (mlegner) [inline crates/walrus-service/src/node/shard_sync.rs]: Can this be removed?
```suggestion
```

File: crates/walrus-service/src/test_utils.rs [hunk 22]
```diff
@@ -250,6 +251,8 @@ pub struct SimStorageNodeHandle {
     pub cancel_token: CancellationToken,
     /// The wrapped simulator node id.
     pub node_id: sui_simulator::task::NodeId,
+    /// The storage capability object of the node.
+    pub storage_capability: Option<StorageNodeCap>,
```
Comment (mlegner) [inline crates/walrus-service/src/test_utils.rs]: Nit: Can we call this `storage_node_capability`?

File: crates/walrus-service/src/test_utils.rs [hunk 23]
```diff
@@ -1649,6 +1680,9 @@ pub mod test_cluster {
         node::{committee::DefaultNodeServiceFactory, contract_service::SuiSystemContractService},
     };
 
+    /// The weight of each storage node in the test cluster.
+    pub const FROSTER_PER_NODE_WEIGHT: u64 = 1_000_000;
```
Comment (mlegner) [inline crates/walrus-service/src/test_utils.rs]: Nit: Should this be `FROST`?
```suggestion
    pub const FROST_PER_NODE_WEIGHT: u64 = 1_000_000;
```

File: crates/walrus-service/src/node/storage.rs [hunk 24]
```diff
@@ -52,13 +55,48 @@ pub(crate) use shard::{ShardStatus, ShardStorage};
...
+}
+
+impl Display for NodeStatus {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let display_string = match self {
+            NodeStatus::Active => "Active".to_string(),
+            NodeStatus::RecoveryCatchUp => "RecoveryCatchUp".to_string(),
+            NodeStatus::RecoveryInProgress(epoch) => format!("RecoveryInProgress ({epoch})"),
+        };
+        write!(f, "{}", display_string)
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage.rs]: Nit: This version seems simpler and clearer to me:
```suggestion
        match self {
            NodeStatus::Active => write!(f, "Active"),
            NodeStatus::RecoveryCatchUp => write!(f, "RecoveryCatchUp"),
            NodeStatus::RecoveryInProgress(epoch) => write!(f, "RecoveryInProgress ({epoch})"),
        }
```

File: crates/walrus-service/src/node/storage.rs (hunk 12 above)
Comment (mlegner) [reply crates/walrus-service/src/node/storage.rs]: Nit:
```suggestion
```

File: crates/walrus-service/src/node/node_recovery.rs [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+            if recovering_epoch == self.node.current_epoch() {
+                return self.start_node_recovery(self.node.current_epoch());
+            } else {
+                assert!(recovering_epoch < self.node.current_epoch());
+                tracing::warn!(
+                    recovering_epoch = recovering_epoch,
+                    current_epoch = self.node.current_epoch(),
+                    "recovery epoch mismatch; skip recovery restart; next epoch change start event
+                    will bring node to the latest state"
+                );
```
Comment (mlegner) [inline crates/walrus-service/src/node/node_recovery.rs:138]: Important: Don't we have to set the status to `RecoveryCatchUp` for this to be true?

File: crates/walrus-service/src/node/node_recovery.rs [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+            }
+        });
+        *locked_task_handle = Some(task_handle);
+
+        Ok(())
+    }
+
+    /// Restarts any in progress recovery.
+    pub fn restart_recovery(&self) -> Result<(), TypedStoreError> {
+        if let crate::node::NodeStatus::RecoveryInProgress(recovering_epoch) =
```
Comment (mlegner) [inline crates/walrus-service/src/node/node_recovery.rs]: Nit:
```suggestion
        if let NodeStatus::RecoveryInProgress(recovering_epoch) =
```

File: crates/walrus-service/src/node/node_recovery.rs [hunk 27]
```diff
@@ -0,0 +1,155 @@
...
+                                        blob_id, err,
+                                    );
+                                }
+                            }
+                        }
+                        Err(err) => {
+                            tracing::error!(error = ?err, "failed to read certified blob");
+                        }
+                    }
+                }
```
Comment (mlegner) [inline crates/walrus-service/src/node/node_recovery.rs]: How about this version that reduces the indentation by two levels?
```suggestion
                for (blob_id, blob_info) in node
                    .storage
                    .certified_blob_info_iter_before_epoch(epoch)
                    .filter_map(|blob_result| {
                        blob_result.inspect_err(|error| {
                            tracing::error!(?error, "failed to read certified blob")
                        })
                        .ok()
                    })
                {
                    if !blob_info.is_certified(epoch) {
                        // Skip blobs that are not certified in the given epoch. This
                        // includes blobs that are invalid or expired.
                        tracing::debug!(%blob_id, %epoch, "skip non-certified blob");
                        continue;
                    }

                    if let Ok(stored_at_all_shards) = node.storage.is_stored_at_all_shards(&blob_id)
                    {
                        if stored_at_all_shards {
                            tracing::debug!(
                                %blob_id,%epoch,"blob is stored at all shards; skip recovery"
                            );
                            continue;
                        }
                    } else {
                        tracing::warn!(
                            %blob_id,
                            "failed to check if blob is stored at all shards; start blob sync"
                        );
                    }

                    tracing::debug!(%blob_id, "start recovery sync for blob");

                    // TODO: rate limit start sync to avoid OOM.
                    let start_sync_result = blob_sync_handler
                        .start_sync(
                            blob_id,
                            blob_info.initial_certified_epoch().expect(
                                "certified blob should have an initial certified epoch set",
                            ),
                            None,
                            start,
                        )
                        .await;
                    match start_sync_result {
                        Ok(notify) => {
                            all_blob_syncs.push(notify);
                        }
                        Err(err) => {
                            // The only place where start_sync can fail is when marking the
                            // event complete, which is not applicable here since the there
                            // is no event associated with the recovery task.
                            panic!(
                                "failed to start recovery sync for blob {}: {}",
                                blob_id, err,
                            );
                        }
                    }
```

File: crates/walrus-service/src/node/committee.rs (hunk 2 above)
Comment (halfprice) [reply crates/walrus-service/src/node/committee.rs]: Done

File: crates/walrus-sdk/src/client.rs (hunk 16 above)
Comment (halfprice) [reply crates/walrus-sdk/src/client.rs]: Good point! I didn't read where the `server_health_info` is used carefully...

File: crates/walrus-service/src/node/blob_sync.rs (hunk 19 above)
Comment (halfprice) [reply crates/walrus-service/src/node/blob_sync.rs]: Amazing catch! Add a field to be explicit about whether the sync is cancelled.

File: crates/walrus-service/src/node/metrics.rs (hunk 20 above)
Comment (halfprice) [reply crates/walrus-service/src/node/metrics.rs:97]: Yeah, I'm aware of that and I suspect using field to indicate status also hard to observe in dashboard, since there will me multiple lines jumping between 0 and 1. So I guess we keep it as is? Let me know what do you prefer.

File: crates/walrus-service/src/test_utils.rs (hunk 22 above)
Comment (halfprice) [reply crates/walrus-service/src/test_utils.rs]: Sure. Done

File: crates/walrus-service/src/node/node_recovery.rs (hunk 25 above)
Comment (halfprice) [reply crates/walrus-service/src/node/node_recovery.rs:138]: And of course you are right! I thought about this issue pretty late last night, and only added the branch without thinking through it. I think what needed is a test to do repeated crash-recovery. I'll create one tomorrow.

File: crates/walrus-service/src/node.rs (hunk 15 above)
Comment (halfprice) [reply crates/walrus-service/src/node.rs]: Done

File: crates/walrus-service/src/node/metrics.rs (hunk 20 above)
Comment (mlegner) [reply crates/walrus-service/src/node/metrics.rs:97]: Yes, let's keep it as it is. 👍

File: crates/walrus-service/src/client.rs [hunk 28]
```diff
@@ -1148,18 +1148,32 @@ where
...
+                tracing::warn!(
+                    %error,
+                    "an error occurred during the current operation, \
+                    which may be caused by epoch change; refreshing the committees and retrying"
+                );
+                client.refresh_committees(current_epoch).await?;
+            }
+            result => return result,
         }
-        result => result,
```
Comment (mlegner) [inline crates/walrus-service/src/client.rs]: Question: What motivated these changes? AFAIU, the `request_rate_config.max_retries` is intended for individual requests to the storage nodes, not a complete operation. Also, if the error occurs due to an epoch change, retrying once should resolve it.

So my suggestion would be to revert these changes. cc @giac-mysten

Comment (giac-mysten) [reply crates/walrus-service/src/client.rs]: yes, this retry mechanism is in place only to refresh the committee if the client fails because of epoch change. 

The only reason why it may make sense to try multiple times is if the client tries to refresh, but fails for so long that another epoch change occurs -- i don't think we should consider this case though 😅

File: crates/walrus-service/src/node/committee/committee_service.rs [hunk 29]
```diff
@@ -611,6 +604,59 @@ where
...
+            )
+            .await
+            .map_err(BeginCommitteeChangeError::AllServicesFailed)?;
+
+            let mut services = self
+                .inner
+                .services
+                .lock()
+                .expect("thread did not panic with mutex");
+            services.extend(new_services);
```
Comment (giac-mysten) [inline crates/walrus-service/src/node/committee/committee_service.rs]: Suggestion: this code is duplicated twice. separate into a function `self.extend_services_from_committee` or similar?

File: crates/walrus-service/src/node/committee/committee_service.rs [hunk 30]
```diff
@@ -392,14 +392,7 @@ where
...
-            committees.current_committee(),
-            &encoding_config,
-        )
-        .await?;
-        add_members_from_committee(
-            &mut services,
+        let services = create_services_from_committee(
             &mut service_factory,
             committees.current_committee(),
             &encoding_config,
```
Comment (giac-mysten) [inline crates/walrus-service/src/node/committee/committee_service.rs:417]: Question: was this a bug where the same services were added twice?

File: crates/walrus-service/src/node/committee/committee_service.rs [hunk 31]
```diff
@@ -611,6 +604,59 @@ where
...
+                .services
+                .lock()
+                .expect("thread did not panic with mutex");
+            services.extend(new_services);
+        }
+
+        self.inner.committee_tracker.send_if_modified(|tracker| {
+            tracker.update_active_committees(latest);
+            true
+        });
```
Comment (giac-mysten) [inline crates/walrus-service/src/node/committee/committee_service.rs]: Question: are we absolutely sure that the committe has been modified?

Suggestion: if we are sure, then
```suggestion
        self.inner.committee_tracker.send_modfiy(|tracker| {
            tracker.update_active_committees(latest);
        });
```

File: crates/walrus-service/src/node/committee/committee_service.rs [hunk 32]
```diff
@@ -611,6 +604,59 @@ where
     fn end_committee_change(&self, epoch: Epoch) -> Result<(), EndCommitteeChangeError> {
         self.end_committee_change_to(epoch)
     }
+
+    async fn begin_committee_change_to_latest_committee(
```
Comment (giac-mysten) [inline crates/walrus-service/src/node/committee/committee_service.rs:627]: Question/imporant: since were are completely replacing the committee with the current one fetched from chain, should we also remove the old entries in `services`, in addition to extending with the new ones? similar to `end_commitee_change_to`?

File: crates/walrus-service/src/node/committee/committee_service.rs (hunk 13 above)
Comment (giac-mysten) [reply crates/walrus-service/src/node/committee/committee_service.rs]: Thanks @halfprice ! left just a few nits and questions.

File: crates/walrus-service/src/node/committee/committee_service.rs (hunk 32 above)
Comment (giac-mysten) [reply crates/walrus-service/src/node/committee/committee_service.rs:627]: maybe i'm wrong, but seems to me that the `end_committee_change` will only remove the services for the previous committee (newly updated), instead of all the ones that may be left from the original (very old) committee.

File: crates/walrus-service/src/node/committee/committee_service.rs (hunk 30 above)
Comment (halfprice) [reply crates/walrus-service/src/node/committee/committee_service.rs:417]: Yeah, `add_members_from_committee` is already called inside `create_services_from_committee`: https://github.com/MystenLabs/walrus/blob/9e74f03a72791244ed79da22652e893e84e388e9/crates/walrus-service/src/node/committee/committee_service.rs#L691

File: crates/walrus-service/src/client.rs (hunk 28 above)
Comment (halfprice) [reply crates/walrus-service/src/client.rs]: The reason for this is exactly what @giac-mysten mentioned: after retry, before finishing the retry, another epoch change is triggered. This is primarily due to that in the test, I set epoch duration to 30 seconds, and blob uploads also takes around 30 seconds (simulated time) to finish.

In the latest commit, I changed the workload to use very small blob size, which helps with this situation. So I reverted the repeated retry. Let's see if CI is ok.

File: crates/walrus-service/src/node/committee/committee_service.rs (hunk 29 above)
Comment (halfprice) [reply crates/walrus-service/src/node/committee/committee_service.rs]: Done

File: crates/walrus-service/src/node/committee/committee_service.rs (hunk 32 above)
Comment (halfprice) [reply crates/walrus-service/src/node/committee/committee_service.rs:627]: Good call out. I seems to recall that when @jpcsmith was talking about committee change, he mentioned that for any ongoing RPCs, removing the service would cause the RPC to hang. However, I just test it that before creating new services, I clear the existing services since we will just use the latest one. The test seems ok.

Comment (giac-mysten) [reply crates/walrus-service/src/node/committee/committee_service.rs:627]: I see, where do yuo remove the old services?

Comment (halfprice) [reply crates/walrus-service/src/node/committee/committee_service.rs:627]: Line 638 below, where I simply clear the services map.

Comment (halfprice) [reply crates/walrus-service/src/node/committee/committee_service.rs:627]: Hi @giac-mysten , I'm going to merge the PR for now, but feel free to let me know if you have other comments. Thanks for pointing out the issue regarding removing services.
